### DISCUSSION TITLE
关于ER图自动布局实现方法

### DISCUSSION THREAD

**Discussion by world56** [antvis/X6#2495] [Q&A] [answered]
import { ERLayout } from "@antv/layout/lib/layout/er/index";

ERLayout这个方法没有任何使用教程。X6官方只提供了GridLayout的使用教程。

X6 ER DEMO，全部都是手动写的position，x:260,y:130 属性
<img width="566" alt="image" src="https://user-images.githubusercontent.com/37354312/184058381-04be9123-6671-4eeb-b506-a23492a7dcbd.png">

有大佬知道ER图怎么实现自动布局吗？或者ERLayout的使用方法？

**附：**
ERLayout源码地址：
https://github.com/antvis/layout/blob/master/src/layout/er/index.ts

**Comment by world56** [comment]
如果去掉demo 所有手写的固定位置属性   "position": {"x": 24,"y": 150}     就全部集中在一起了。

<img width="309" alt="image" src="https://user-images.githubusercontent.com/37354312/184059852-d92938b8-8ce2-47a1-ab48-9a13cf7daae7.png">

在线地址：
https://codesandbox.io/s/trusting-cohen-mwh9rn?file=/index.ts

**Comment by zmcode** [comment]
这个ERlayout内部使用了GridLayout和force布局, 你直接使用它内部源码的布局会很卡的, 它内部的force没有做静态布局的处理, 先说说你的具体想画出啥的页面

**Reply by world56** [reply]
🙏感谢回答，如果不设置 “position” 属性，整个ER图 全部会柔在一起重叠。

能让他自动布局分散开吗？

因为之前用G6都没有去主动设置元素位置（position）

**Reply by zmcode** [reply]
例子中, position其实就是这个节点的对应坐标的 x,y位置, 你去除了肯定都叠在一起了, json中的每个对象其实是每个自定义节点的对应node数据, NODE_WIDTH 就是每个组的宽度和 LINE_HEIGHT是 每行的高度, 你遍历你的数据格式的时候, 依次递增NODE_WIDTH 就可以了, 在加一点的宽度让他们排开, 这个要看你最终要显示出怎么样的效果, G6的布局方案你得改它的源码, 太复杂了, 我看了它的force, 没有用静态布局的, 如果你的节点一多, 它会不断计算位置才会得到最后计算结果的

**Reply by world56** [reply]
有ERLayout的使用教程吗？卡就先卡着吧，凑合一下😭

**Comment by world56** [comment]
手动去算position，主要涉及到回显的问题。

全部有序排列，如果其中一个Node稍微移动了一丢丢，二次在回显，是判断他继续依次排列，还是与其他Node重叠？

所以感觉能用官方的ERLayout方案最好，就初始化执行布局一次就行。卡顿一点儿也无所谓。

**Comment by zmcode** [comment, marked as answer]
https://codesandbox.io/s/quizzical-robinson-mso55e?file=/index.ts
大概是这样使用

**Reply by zmcode** [reply]
我看了下force不会导致卡顿, 最终没有渲染数据, 只是做计算

**Reply by world56** [reply]
> 我下力不会导致卡顿，最终没有渲染数据，只是做计算

牛逼，这个可以推荐写入官方文档。

**Reply by world56** [reply]
> https://codesandbox.io/s/quizzical-robinson-mso55e?file=/index.ts 大概是这样使用

私信我发个你的支付宝帐号，请你喝杯奶茶。🥤

**Reply by zmcode** [reply]
> > https://codesandbox.io/s/quizzical-robinson-mso55e?file=/index.ts 大概是这样使用
> 
> 私信我发个你的支付宝帐号，请你喝杯奶茶。🥤

只是简单的例子而已, 还有很多复杂的配置要看那个布局文档了, 能帮到你就行,我自己也进步的

**Reply by world56** [reply]
> > > https://codesandbox.io/s/quizzical-robinson-mso55e?file=/index.ts 大概是这样使用
> > 
> > 
> > 私信我发个你的支付宝帐号，请你喝杯奶茶。🥤
> 
> 只是简单的例子而已, 还有很多复杂的配置要看那个布局文档了, 能帮到你就行,我自己也进步的

感谢大佬🙏

**Reply by 3403784285-lucky** [reply]
啊啊啊，为什么这个在线我点进去没有了了，还报错
![image](https://github.com/user-attachments/assets/9f88c210-27a3-4662-9af8-7e0add3ee869)

**Comment by tangzhexu** [comment]
请问下如何禁用调ER图的手动创建连接功能

**Comment by 3403784285-lucky** [comment]
现在这个布局是不是没有了
